File: home-work/affine.m
clear;close all;clc
im = imread("../example-grey-small.png");
% imshow(im);

theta = pi / 6;
rot_mat = [cos(theta), -sin(theta), 0;
    sin(theta), cos(theta), 0;
    0, 0, 1];
shear_mat = [1, 0.3, 0; 0, 1, 0; 0, 0, 1];
affine_mat = rot_mat * shear_mat;
[raw_x, raw_y] = size(im);

maxx = 0; maxy = 0; minx = 0; miny = 0;
for x = [1 raw_x]
    for y = [1 raw_y]
        cood = affine_mat * [x; y; 1];
        maxx = max(cood(1), maxx);
        minx = min(cood(1), minx);
        maxy = max(cood(2), maxy);
        miny = min(cood(2), miny);
    end
end

shift_mat = [1, 0, abs(min(0, minx)); 0, 1, abs(min(0, miny)); 0, 0, 1];
T = shift_mat * rot_mat * shear_mat;
inv_T = inv(T);
maxx2 = 0; maxy2 = 0; minx2 = 0; miny2 = 0;
for x = [1 raw_x]
    for y = [1 raw_y]
        cood2 = T * [x; y; 1];
        disp(cood2)
        maxx2 = max(cood2(1), maxx2);
        minx2 = min(cood2(1), minx2);
        maxy2 = max(cood2(2), maxy2);
        miny2 = min(cood2(2), miny2);
    end
end

size_x = ceil(maxx2); size_y = ceil(maxy2);
new_im = zeros(size_x, size_y);
for x=1:size_x
    for y=1:size_y
        raw_pos = inv_T * [x;y;1];
        sx = raw_pos(1); sy = raw_pos(2);
        if sx < 1  || sy < 1 || sx > raw_x || sy > raw_y
            new_im(x,y) = 255;
            continue
        end
        % insert value
        %         a = im(floor(sx),floor(sy)) * (ceil(sy) - sy) + ...
        %             im(floor(sx),ceil(sy)) * (sy - ceil(sy));
        %         b = im(ceil(sx),floor(sy)) * (ceil(sy) - sy) + ...
        %             im(ceil(sx),ceil(sy)) * (sy - floor(sy));
        %         target = a * (ceil(sx)-sx) + b * (sx-floor(sx));
        alpha = im(floor(sx), floor(sy)); beta = im(floor(sx), ceil(sy));
        galma = im(ceil(sx), floor(sy)); theta = im(ceil(sx), ceil(sy));
        p = sy - floor(sy); q = sx - floor(sx);
        target = (alpha*(1-p)+beta*p) * (1-q) + ...
            (galma*(1-p)+theta*p) * (q);
        new_im(x,y)=target;
    end
end
imshow(uint8(new_im))

File: home-work/affine.py
from PIL import Image
import numpy as np
from matplotlib import pyplot as plt
import math
from itertools import product

im_raw = Image.open("./example-grey-small.png")
im = np.array(im_raw)
size = im.shape
rotaion_angle = math.pi / 6
rotaion_mat = np.matrix([[math.cos(rotaion_angle), -math.sin(rotaion_angle), 0],
                        [math.sin(rotaion_angle), math.cos(rotaion_angle), 0],
                        [0, 0, 1]])
shear_mat = np.matrix([[1, 0.3, 0], [0, 1, 0], [0, 0, 1]])
shift_mat = np.matrix([[1, 0, 123], [0, 1, 0], [0, 0, 1]])
affine_mat = rotaion_mat * shear_mat * shift_mat
revert_affine_mat = affine_mat.I
print(size)
print(affine_mat)

max_pos = [0, 0]
for i, j in product([0, size[0] - 1], [0, size[1] - 1]):
    trans_pos = np.array(affine_mat * np.matrix([i, j, 1]).transpose())
    if trans_pos[0][0] > max_pos[0]:
        max_pos[0] = trans_pos[0][0]
    if trans_pos[1][0] > max_pos[1]:
        max_pos[1] = trans_pos[1][0]

max_size = tuple(map(lambda x: math.ceil(x)+1, max_pos))[:2]
new_im = np.zeros(max_size)

cnt = 0
for i in range(max_size[0]):
    for j in range(max_size[1]):
        source_pos = np.array(revert_affine_mat * np.matrix([i, j, 1]).transpose())
        source_x = source_pos[0][0]
        source_y = source_pos[1][0]
        if source_x <0 or source_x > size[0] - 1 or source_x < 0 or source_y > size[1] - 1:
            new_im[i][j] = 255
            continue
        # 插值
        a = im[math.floor(source_x)][math.floor(source_y)] * (math.ceil(source_y) - source_y) + \
            im[math.floor(source_x)][math.ceil(source_y)] * (source_y - math.floor(source_y))
        b = im[math.ceil(source_x)][math.floor(source_y)] * (math.ceil(source_y) - source_y) + \
            im[math.ceil(source_x)][math.ceil(source_y)] * (source_y - math.floor(source_y))
        target = a * (math.ceil(source_x) - source_x) + b * (source_x - math.floor(source_x))
        new_im[i][j] = int(target)
new_im = Image.fromarray(np.uint8(new_im))
new_im.show()